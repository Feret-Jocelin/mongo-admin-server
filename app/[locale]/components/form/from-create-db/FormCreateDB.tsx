'use client';
import { useState } from 'react';
import { useI18n } from '@/shared/locales/clients';
import { useRouter } from 'next/navigation';

import GenericInput from '@components/ui/inputs/generic/GenericInput';
import GenericButton from '@components/ui/button/GenericButton';

import styles from './form-create-db.module.scss';
import GenericModal from '../../modal/generic/GenericModal';
import axios from 'axios';

interface CreateDBModalProps {
  open: boolean;
  onConfirm: () => void;
  onClose: () => void;
}

export default function FormCreateDB({
  open,
  onConfirm,
  onClose,
}: CreateDBModalProps) {
  /* Static Data */
  const router = useRouter();
  const t = useI18n();
  /* Methods */
  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const dbCreateData = {
      databaseName: nameDB,
      collectionName: nameFirstCollection
    }

    if (dbCreateData.databaseName === '' || dbCreateData.collectionName === '') {
      setError('Veuillez remplir tous les champs.');
      return;
    }

    setError('');

    const url = `/api/database`;
    axios
      .post(url, dbCreateData)
      .then(response =>{
        console.log(response);
        router.push(`/dashboard/${dbCreateData.databaseName}`)
      })
      .catch((error) => {
        if (error.response) {
          console.log(error.response);
          console.log("server responded");
          if(error.response.status === 500){
            setError(t('formCreateDB.error500'))
          } else if(error.response.status === 409) {
            setError(t('formCreateDB.error409'))
          } else {
            console.log(error.response);
          }
        } else if (error.request) {
          console.log("network error");
        } else {
          console.log(error);
        }
      });

  };
  
  const handleNameDBChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setNameDB(event.target.value);
  };
  const handleNameFirstCollectionChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setNameFirstCollection(event.target.value);
  };

  /* Local Data */
  const [nameDB, setNameDB] = useState<string>('');
  const [nameFirstCollection, setNameFirstCollection] = useState<string>('');
  const [error, setError] = useState<string>('');
  const isSubmitButtonDisabled = nameDB === '' || nameFirstCollection === '';

  return (
    <div className={styles.form_create}>
      <GenericModal
        open={open}
        title={t('formCreateDB.title')}
        withButton={false}
        onClose={onClose}
        onValidate={onConfirm}
      >
        <div className={styles.formContainer}>
          <div className={styles.container}>
            <form className={styles.form} onSubmit={handleSubmit}>
              <div className={styles.formGroup}>
                <div className={styles.formInput}>
                  <GenericInput
                    type="text"
                    label={t('formCreateDB.nameDB')}
                    value={nameDB}
                    placeholder={t('formCreateDB.nameDB')}
                    error={error}
                    onChange={handleNameDBChange}
                  />
                </div>
                <div className={styles.formInput}>
                  <GenericInput
                    type="text"
                    label={t('formCreateDB.nameFirstCollection')}
                    value={nameFirstCollection}
                    placeholder={t('formCreateDB.nameFirstCollection')}
                    error={error}
                    onChange={handleNameFirstCollectionChange}
                    />
                </div>
              </div>
              <div className={styles.formButton}>
                <GenericButton center outline onClick={onClose}>
                  cancel
                </GenericButton>
                <GenericButton center disabled={isSubmitButtonDisabled} type="submit">
                  {t('formCreateDB.title')}
                </GenericButton>
              </div>
            </form>
          </div>
        </div>
      </GenericModal>
    </div>
  );
}
