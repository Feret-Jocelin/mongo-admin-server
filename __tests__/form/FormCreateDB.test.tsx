/* eslint-disable no-undef */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { NextIntlClientProvider } from 'next-intl';
import { store } from '@/store/store';
import { Provider } from 'react-redux';
import FormCreateDB from '@/app/[locale]/components/form/from-create-db/FormCreateDB';

jest.mock('next-intl/client', () => ({
  usePathname() {
    return '';
  },
  useRouter() {
    return {
      push: () => jest.fn(),
      replace: () => jest.fn(),
    };
  },
}));

const locale = 'fr';
const message = require(`../../shared/locales/${locale}.json`)

describe("FormCreateDB", () => {
  it('renders', () => {
    const handleClose = jest.fn();
    const { container } = render(
      <Provider store={store}>
        <NextIntlClientProvider locale={locale} messages={message}>
          <FormCreateDB 
            open={true}
            onClose={handleClose}
          />
        </NextIntlClientProvider>
      </Provider>
    );
    const title = screen.getByText('Créer une base de données');
    expect(title).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  })
})