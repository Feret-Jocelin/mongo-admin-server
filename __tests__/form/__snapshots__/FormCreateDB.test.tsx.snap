// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCreateDB renders 1`] = `
<div>
  <dialog
    class="modal"
  >
    <section
      class="modal__header"
    >
      <div
        class="modal__header--icon"
      >
        <div
          class="undefined pointer-cursor"
        />
      </div>
      <h2>
        Créer une base de données
      </h2>
    </section>
    <section
      class="modal__container"
    >
      <form
        class="create-database-form"
      >
        <div
          class="input-container"
        >
          <label
            class="label"
          >
            Base de données
          </label>
          <input
            class="input "
            placeholder="Base de données"
            type="text"
            value=""
          />
        </div>
        <div
          class="input-container"
        >
          <label
            class="label"
          >
            Collections
          </label>
          <input
            class="input "
            placeholder="Collections"
            type="text"
            value=""
          />
        </div>
      </form>
    </section>
    <section
      class="modal__footer"
    >
      <button
        class="button full outline center"
        type="button"
      >
        Annuler
      </button>
      <button
        class="button full center"
        disabled=""
        type="button"
      >
        Créer
      </button>
    </section>
  </dialog>
</div>
`;
